### PR TITLE
Smart menu

### PR DESCRIPTION
**Pull request by stefanscholz**
This pull request contains an early version for the smart menu feature (#137) as well as some foundational work for the menu bar #136.

### CODE REVIEW

File: scss/boost_union/post.scss [hunk 1]
```diff
@@ -917,3 +914,994 @@ body.pagelayout-login:not(.loginbackgroundimage) #footnote {
...
+            height: 42px;
+            border-bottom: 0;
+            background-color: #f8f9fa;
+            z-index: 1031;
+            /* The position of the side drawer has been adjusted slightly below to accommodate the top header. */
+            ~ .drawer.drawer-left,
+            ~ .drawer.drawer-right {
+                top: 100px;
+            }
+           /* The position of the main page has been adjusted slightly below to accommodate the top header. */
```
Comment (lucaboesch) [inline scss/boost_union/post.scss:960]: Just one space more to the right please.

File: scss/boost_union/post.scss [hunk 2]
```diff
@@ -917,3 +914,994 @@ body.pagelayout-login:not(.loginbackgroundimage) #footnote {
...
+                    }
+                    /* The submenu dropdown is hidden before the menu is clicked */
+                    ul {
+                        display: none;
+                    }
+                    /* The position relative is added for the dropdown menu */
+                    &.show {
+                        position: relative;
+                        /* The dropdown submenu style is added for the header menu */
+                        /* Height added to control the menu with scrollbard, border, border radius, background color, padding and its position styles are
```
Comment (lucaboesch) [inline scss/boost_union/post.scss:1260]: Do break the row earlier not to exceed 132 characters in length.

File: scss/boost_union/post.scss [hunk 3]
```diff
@@ -917,3 +914,994 @@ body.pagelayout-login:not(.loginbackgroundimage) #footnote {
...
+                        /* Footer bottom dropdown menu width, height and removed margin, position to stay the dropup menu */
+                        .dropdown-menu {
+                            width: auto;
+                            margin-top: 0;
+                            margin-bottom: 0;
+                            top: auto;
+                            bottom: 100%;
+                            left: auto;
+                            right: 0;
+                            /* The footer bottom dropdown submenu min-width, max-height, overflow and position to show the menus above the footer */
```
Comment (lucaboesch) [inline scss/boost_union/post.scss:1667]: Do break the row to multiple lines and make sure you do not extend 132 characters of line length.

File: scss/boost_union/post.scss [hunk 4]
```diff
@@ -917,3 +914,994 @@ body.pagelayout-login:not(.loginbackgroundimage) #footnote {
...
+
+/* Footer Menu */
+#page-wrapper #page-footer {
+    .navbar {
+        /* Footer bottom menu */
+        /* Padding space removed around the bottom menu block, box-shadow and z-index
+        to view the bottom menu at the top added*/
+        &.boost-union-bottom-menu {
+            padding: 0;
+            box-shadow: 0px 0px 6px 0 rgba(0, 0, 0, 0.1);
```
Comment (lucaboesch) [inline scss/boost_union/post.scss:1631]: 0px isn't valid use
`box-shadow: 0 0 6px 0 rgba(0, 0, 0, 0.1);`

File: scss/boost_union/post.scss (hunk 1 above)
Comment (abias) [reply scss/boost_union/post.scss]: Thanks, Luca.
I fixed this in a review commit now.

File: scss/boost_union/post.scss (hunk 2 above)
Comment (abias) [reply scss/boost_union/post.scss]: Thanks, Luca.
I fixed this in a review commit now.

File: scss/boost_union/post.scss (hunk 3 above)
Comment (abias) [reply scss/boost_union/post.scss]: Thanks, Luca.
I fixed this in a review commit now.

**Comment by wiebkemueller-hsh** [comment]
Hi, 

during functional testing of the smart menu feature in our local testing environment, we made the following observations:

Specification of the local testing environment
PHP 8.1.20
memory_limit = 128MB > raised to 256MB
memory_limit = 256M in php.ini
DB Postgres 12.11 (including a database dump with real data from our live system)
Moodle 4.1.4 (Build: 20230612)
Boost Union:
First try with this PR#300 which was added to v4.1-r7 and then merged/rebased with v4.1-r8
and in a 2nd try the  https://github.com/bdecentgmbh/moodle-theme_boost_union/tree/smart_menu directly to a v4.1-r7

**1. A smart menu item "static" and "heading" cannot be saved.** 
The entry form won't let us save the dialogue, but there is no error message telling us what is wrong. 

To reproduce:

type static URL
create new smart menu, location main navigation, type list, mode submenu
add new item, type static URL (our example /local/hsh/request.php)
try to save

type heading
create new smart menu, location main navigation, type list, mode submenu
add new item, type heading
try to save

**2. Fallback "extra menu item" leads to system crash**

Resulting out of the behaviour described in 1 we tried to implement a fallback which made the whole system crash. We put an extra menu item in the main navigation via the theme settings, that links directly to the URL we wanted to link to in the first place. We had used this functionality before for the time being to provide quick access to the course request form. The idea was to combine this with the smart menu that would then sit next to the custom menu item in the case the static link would take some more time.

To reproduce (see also screen 3):
Go to admin/settings.php?section=themesettings
add URL in custommenuitems (our example is Kurs beantragen|/local/hsh/request.php|Nicht für Studierende)

After this was done the session was destroyed and the user (admin) was logged out. Trying to login and reuse the system constantly reproduces the following errors, see also screen 1 and sometimes screen 2.

Warning: session_start(): Failed to decode session object. Session has been destroyed in /var/www/html/lib/classes/session/handler.php on line 45
Error
error/An error occurred whilst communicating with the server
Debug info:
Error code: An error occurred whilst communicating with the server
$a contents:
Stack trace:

line 159 of /lib/classes/session/manager.php: core\\session\\exception thrown
line 139 of /lib/classes/session/manager.php: call to core\\session\\manager::start_session()
line 842 of /lib/setup.php: call to core\\session\\manager::start()
line 79 of /config.php: cal

We hope the given screenshots and error message help to identify the problem. I tried out the smart menu in https://m41.bdecent.io/ before and adding headings and URLs as static links to the menu worked out fine, so we are really keen on finding out what the problem could be. Are there extra menu items set? I didn’t want to take a risk and crash that system too, so I didn’t add one.

Regards
Wiebke (from HsH)


Screen 1
grafik


Screen 2
grafik


Screen 3
grafik

**Comment by abias** [comment]
I just pushed two commits with my latest review results. They do not change the contributed functionality, but help to integrate the patch into the existing codebase.

**Comment by abias** [comment]
Hi @prasanna-lmsace ,

thank you very much for pushing some very important fixes and improvements to the smart_menu branch which is the basis for this PR.

As I have written above, I am still in the middle of the code review of this really large piece of code. And I have added a "Review changes" commit on top of your developments which polishes your code in many ways and helps to integrate it into the existing codebase.

However, I had to see that you did not rebase your latest changes properly onto the codebase which was already there in the smart_menus branch. You somehow merged my changes into your codebase and then pushed it.

This makes it really hard for me to understand what you have really changed recently.

Please look at this image:

grafik

The commits which I have stroke through must not be there. You should just add additional commits, ideally one commit per fixed issue and improvement.

After speaking to @stefanscholz , I have now foce-pushed my latest reviewed codebase again and have thus reverted your latest changes. I would like to ask you to rebase your work properly and to re-push the commits into the branch. 

To help you to do this, I will not push any review changes until your commits are there again.

Thank you for your cooperation :)

**Comment by abias** [comment]
Thank you @prasanna-lmsace very much for pushing proper commits to the branch now, I highly appreciate your effort!

**Comment by abias** [comment]
I rebased the PR onto latest master where a tiny change was added for the flavours overview table to align both the flavours and smart menu class structure

**Review by abias — APPROVED** [review]

**Comment by abias** [comment]
Thank you all for working on and commenting in this PR!

I will provisionally merge this now into the master branch to allow the remaining minor issues which have been discovered internally to be handled within dediated follow-up Github issues over the next days.

An official release on moodle.org/plugins which includes the smart menus will follow as soon as we have burnt the list of follow-up issues down.

Cheers,
Alex

**Pull request merged**